encodings/fsst/src/canonical.rs

```diff
@@ -1,56 +1,46 @@
-use arrow_array::builder::GenericByteBuilder;
-use arrow_array::types::BinaryType;
-use fsst::Symbol;
-use vortex::array::VarBinArray;
-use vortex::arrow::FromArrowArray;
-use vortex::validity::ArrayValidity;
-use vortex::{ArrayDType, Canonical, IntoCanonical};
+use vortex::array::{PrimitiveArray, VarBinArray};
+use vortex::validity::Validity;
+use vortex::{ArrayDType, Canonical, IntoArray, IntoCanonical};
 use vortex_error::VortexResult;
 
 use crate::FSSTArray;
 
 impl IntoCanonical for FSSTArray {
     fn into_canonical(self) -> VortexResult<Canonical> {
         self.with_decompressor(|decompressor| {
-            // Note: the maximum amount of decompressed space for an FSST array is 8 * n_elements,
-            // as each code can expand into a symbol of 1-8 bytes.
-            let max_items = self.len();
-            let max_bytes = self.codes().nbytes() * size_of::<Symbol>();
-
-            // Create the target Arrow binary array
-            // TODO(aduffy): switch to BinaryView when PR https://github.com/spiraldb/vortex/pull/476 merges
-            let mut builder = GenericByteBuilder::<BinaryType>::with_capacity(max_items, max_bytes);
-
-            // TODO(aduffy): add decompression functions that support writing directly into and output buffer.
-            let codes_array = self.codes().into_canonical()?.into_varbin()?;
-
-            // TODO(aduffy): make this loop faster.
-            for idx in 0..self.len() {
-                if !codes_array.is_valid(idx) {
-                    builder.append_null()
-                } else {
-                    let compressed = codes_array.bytes_at(idx)?;
-                    let value = decompressor.decompress(compressed.as_slice());
-                    builder.append_value(value)
-                }
+            let compressed_bytes = VarBinArray::try_from(self.codes())?.bytes().as_primitive();
+
+            // Bulk-decompress the entire array.
+            let uncompressed_bytes =
+                decompressor.decompress(compressed_bytes.maybe_null_slice::<u8>());
+
+            // Convert the uncompressed_lengths into offsets for building a new VarBinArray.
+            let mut offsets: Vec<i32> = Vec::with_capacity(self.len() + 1);
+            let mut offset = 0;
+            offsets.push(offset);
+
+            let uncompressed_lens_array = self
+                .uncompressed_lengths()
+                .into_canonical()?
+                .into_primitive()?;
+            let uncompressed_lens_slice = uncompressed_lens_array.maybe_null_slice::<i32>();
+
+            for len in uncompressed_lens_slice.iter() {
+                offset += len;
+                offsets.push(offset);
             }
```

Comment:
Isn't this a scan? I have a variant of this in ChunkedArray constructor

Reply:
I found the for loop easier to read personally so I just left this as-is